blocks/main_loop.py

```diff
@@ -156,16 +159,17 @@ def _check_finish_training(self):
         # the iteration the corresponding log record can be found only in
         # the previous row.
         if (self.log.current_row.training_finish_requested or
-                self.log.current_row.keyboard_interrupt_received or
-                self.log.previous_row.keyboard_interrupt_received):
+                self.log.current_row.interrupt_received or
+                self.log.previous_row.interrupt_received):
             raise TrainingFinish
 
-    def _handle_keyboard_interrupt(self, signal_number, frame):
+    def _handle_interrupt(self, signal_number, frame):
         # After receiving a first keyboard interrupt signal,
         # ignore all following ones.
```

Comment:
Right, that explains why you it didn't catch it on the second time, because you need to press it 3 times: The first time `_handle_interrupt` gets called and the signals get restored. The second time, you cause a `KeyboardInterrupt` which brings you straight to the `finally` block that calls `self._run_extensions('after_training')`. The third time actually kills the program.

Maybe this would be good behavior:
* On SIGINT, try to finish the epoch. A second SIGINT will try to finish the batch. The third one will kill your program like usual.
* On SIGTERM, we try to finish the batch and then exit.